### PR TITLE
Support finally block in SE

### LINKED ISSUE #5307: Support finally block in SE
Branch to `finally` block in the new Symbolic Execution engine

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5307

Review per commit is suggested to see the two-step evolution. And effect on the UTs.

### CODE REVIEW

File: analyzers/its/expected/akka.net/Akka--netstandard2.0-S2222.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+"issues":  [
+{
+"id":  "S2222",
+"message":  "Unlock this lock along all executions paths of this method.",
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka\Actor\RepointableActorRef.cs",
```
Comment (costin-zaharia-sonarsource) [inline analyzers/its/expected/akka.net/Akka--netstandard2.0-S2222.json:7]: I was hoping for a first TP in an external project. Not yet :)

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/FinallyPoint.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn
+{
+    internal sealed class FinallyPoint
+    {
+        private readonly ControlFlowBranch branch;
+        private readonly int finallyIndex;
+
+        public bool IsFinallyBlock => finallyIndex < branch.FinallyRegions.Length;
+        public int BlockIndex => IsFinallyBlock ? branch.FinallyRegions[finallyIndex].FirstBlockOrdinal : branch.Destination.Ordinal;
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/FinallyPoint.cs:32]: So we now consider the `finally` regions as blocks that appear before any other destination block.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/FinallyPointTest.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+using SonarAnalyzer.UnitTest.Helpers;
+
+namespace SonarAnalyzer.UnitTest.SymbolicExecution.Roslyn
+{
+    [TestClass]
+    public class FinallyPointTest
+    {
+        [TestMethod]
+        public void Constructor_Null_Throws() =>
+            ((Func<FinallyPoint>)(() => new FinallyPoint(null, null))).Should().Throw<ArgumentNullException>().And.ParamName.Should().Be("branch");
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/FinallyPointTest.cs:36]: It throws only if the second parameter is null. I think we should highlight that in the test.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/its/expected/akka.net/Akka--netstandard2.0-S2222.json (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/its/expected/akka.net/Akka--netstandard2.0-S2222.json:7]: Not yet 😒 But this is a bug rule, there's a high chance that the production code already suffered from it in the past and it's therefore fixed. Would be interesting to find TP on Peach one day.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/FinallyPoint.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/FinallyPoint.cs:32]: As a big picture idea, yes. 

(while keeping in mind that the finally region can contain multiple blocks)

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**